sql/parser/type_check.go

```diff
@@ -228,6 +231,63 @@ func typeCheckComparisonOp(op ComparisonOp, dummyLeft, dummyRight Datum) (Datum,
 		dummyLeft.Type(), op, dummyRight.Type())
 }
 
+func typeCheckConditionalExpr(expr *ConditionalExpr) (Datum, error) {
+	switch expr.Operator {
+	case If:
+		if len(expr.Exprs) != 3 {
```

Comment:
Fitting `CaseExpr` into a `ConditionalExpr` interface for type-checking seems awkward as `CaseExpr` has quite a bit of different logic for handling the `when` conditions vs the `when` values. Let me give the `ConditionalExpr` interface a try for the existing conditionals.